### PR TITLE
Fix Method Pointer Update Notifications

### LINKED ISSUE #1029: Missing `executionContext/expressionValuesComputed` After Popping Frame
### General Summary
Consider the following code:
```
Main.mulAdd = 5

main =
    result = Main.mulAdd
    6
```

Now I run it in IDE. Select the node with `Main.mulAdd` expression. Enter it. Leave it. Try to enter it again — fail. It is impossible to enter it again, because IDE invalidates its computed values registry after pushing or popping the stackframe. However, engine does not send `executionContext/expressionValuesComputed` for the said node after popping the frame from the execution context.

Observe the following communication log:
```
{"jsonrpc":"2.0","id":0,"method":"session/initProtocolConnection","params":{"clientId":"e8b12613-752a-44bf-84ef-8c8ec1e3b02c"}}	127	
{"jsonrpc":"2.0","id":0,"result":{"contentRoots":["d1c75f23-bda3-4738-81fc-d9222d193be3"]}}	91	
{"jsonrpc":"2.0","id":1,"method":"search/getSuggestionsDatabase","params":{}}	77	
{"jsonrpc":"2.0","id":1,"result":{"entries":[{"id":47,"suggestion":{"type":"local","externalId":"2d6c6cae-414b-4174-8450-afabd042ae93","module":"Unnamed.Main","name":"result","returnType":"Any","scope":{"start":{"line":2,"character":6},"end":{"line":4,"character":5}}}},{"id":46,"suggestion":{"type":"method","externalId":"abd6952c-ab14-4574-8b8d-a089142c37cf","module":"Unnamed.Main","name":"main","arguments":[{"name":"this","reprType":"Any","isSuspended":false,"hasDefault":false,"defaultValue":null}],"selfType":"here","returnType":"Number"}},{"id":45,"suggestion":{"type":"method","externalId":"2c980296-5cb9-42ac-adea-095409fa7c3a","module":"Unnamed.Main","name":"mulAdd","arguments":[{"name":"this","reprType":"Any","isSuspended":false,"hasDefault":false,"defaultValue":null}],"selfType":"Main","returnType":"Number"}}],"currentVersion":155}}	849	
{"jsonrpc":"2.0","id":2,"method":"capability/acquire","params":{"method":"search/receivesSuggestionsDatabaseUpdates","registerOptions":{}}}	139	
{"jsonrpc":"2.0","id":2,"result":null}	38	
{"jsonrpc":"2.0","id":3,"method":"text/openFile","params":{"path":{"rootId":"d1c75f23-bda3-4738-81fc-d9222d193be3","segments":["src","Main.enso"]}}}	148	
{"jsonrpc":"2.0","id":3,"result":{"writeCapability":{"method":"text/canEdit","registerOptions":{"path":{"rootId":"d1c75f23-bda3-4738-81fc-d9222d193be3","segments":["src","Main.enso"]}}},"content":"Main.mulAdd = 5\n\nmain =\n result = Main.mulAdd\n 6\n\n\n#### METADATA ####\n[[{\"index\":{\"value\":0},\"size\":{\"value\":4}},\"8f6eec81-a516-4dde-be43-07de625a6653\"],[{\"index\":{\"value\":4},\"size\":{\"value\":1}},\"b95addc1-dffd-4622-ba63-4468d083908c\"],[{\"index\":{\"value\":5},\"size\":{\"value\":6}},\"429638bf-fa87-4071-8114-ca700b46fe81\"],[{\"index\":{\"value\":0},\"size\":{\"value\":11}},\"1e0db7d9-0de1-450d-b137-4fa0e2049a35\"],[{\"index\":{\"value\":12},\"size\":{\"value\":1}},\"68d4054d-00d3-49fa-83bd-0807947e28f0\"],[{\"index\":{\"value\":14},\"size\":{\"value\":1}},\"2c980296-5cb9-42ac-adea-095409fa7c3a\"],[{\"index\":{\"value\":0},\"size\":{\"value\":15}},\"f8e67f93-4366-42f9-aa3e-d1d210002b38\"],[{\"index\":{\"value\":17},\"size\":{\"value\":4}},\"b3dbe0d8-4323-4ca2-aacc-14e3b0473740\"],[{\"index\":{\"value\":22},\"size\":{\"value\":1}},\"186f956b-f402-4648-98ce-8b11bea71a53\"],[{\"index\":{\"value\":28},\"size\":{\"value\":6}},\"743ce364-07f6-465f-8640-11fc0cbf306e\"],[{\"index\":{\"value\":35},\"size\":{\"value\":1}},\"612de1cf-9e9e-464b-be5e-7d6ffaa7c694\"],[{\"index\":{\"value\":37},\"size\":{\"value\":4}},\"fecdad1c-9451-4d2b-a7fd-4aad5550de40\"],[{\"index\":{\"value\":41},\"size\":{\"value\":1}},\"27938428-f93b-429e-9094-2fad6d363c83\"],[{\"index\":{\"value\":42},\"size\":{\"value\":6}},\"edf2376f-4edf-4c43-9111-461dc29cef3d\"],[{\"index\":{\"value\":37},\"size\":{\"value\":11}},\"2d6c6cae-414b-4174-8450-afabd042ae93\"],[{\"index\":{\"value\":28},\"size\":{\"value\":20}},\"7ba77f90-2cf9-4402-a3ab-d6f1e6585d51\"],[{\"index\":{\"value\":53},\"size\":{\"value\":1}},\"20f2ff1b-8932-48e5-ba25-b59d15bf7d58\"],[{\"index\":{\"value\":23},\"size\":{\"value\":31}},\"abd6952c-ab14-4574-8b8d-a089142c37cf\"],[{\"index\":{\"value\":17},\"size\":{\"value\":37}},\"0df7d0e1-a95a-47f3-9ea5-b9e54919db2b\"]]\n{\"ide\":{\"node\":{\"20f2ff1b-8932-48e5-ba25-b59d15bf7d58\":{\"position\":{\"vector\":[-495.9664,199.92416]}},\"2d6c6cae-414b-4174-8450-afabd042ae93\":{\"position\":{\"vector\":[-652.0,87.0]}},\"7ba77f90-2cf9-4402-a3ab-d6f1e6585d51\":{\"position\":{\"vector\":[-584.0,115.0]}},\"32dd808a-0633-4648-994c-691653c6e758\":{\"position\":{\"vector\":[-506.0,116.0]}}}}}","currentVersion":"a03427902ca6c19db8721062925a1839133762ddc1498cca25be7b10"}}	2495	
{"jsonrpc":"2.0","id":4,"method":"executionContext/create","params":{}}	71	
{"jsonrpc":"2.0","id":5,"method":"text/applyEdit","params":{"edit":{"edits":[{"range":{"end":{"character":336,"line":9},"start":{"character":0,"line":0}},"text":"Main.mulAdd = 5\n\nmain =\n result = Main.mulAdd\n 6\n\n\n#### METADATA ####\n[[{\"index\":{\"value\":0},\"size\":{\"value\":4}},\"8f6eec81-a516-4dde-be43-07de625a6653\"],[{\"index\":{\"value\":4},\"size\":{\"value\":1}},\"b95addc1-dffd-4622-ba63-4468d083908c\"],[{\"index\":{\"value\":5},\"size\":{\"value\":6}},\"429638bf-fa87-4071-8114-ca700b46fe81\"],[{\"index\":{\"value\":0},\"size\":{\"value\":11}},\"1e0db7d9-0de1-450d-b137-4fa0e2049a35\"],[{\"index\":{\"value\":12},\"size\":{\"value\":1}},\"68d4054d-00d3-49fa-83bd-0807947e28f0\"],[{\"index\":{\"value\":14},\"size\":{\"value\":1}},\"2c980296-5cb9-42ac-adea-095409fa7c3a\"],[{\"index\":{\"value\":0},\"size\":{\"value\":15}},\"f8e67f93-4366-42f9-aa3e-d1d210002b38\"],[{\"index\":{\"value\":17},\"size\":{\"value\":4}},\"b3dbe0d8-4323-4ca2-aacc-14e3b0473740\"],[{\"index\":{\"value\":22},\"size\":{\"value\":1}},\"186f956b-f402-4648-98ce-8b11bea71a53\"],[{\"index\":{\"value\":28},\"size\":{\"value\":6}},\"743ce364-07f6-465f-8640-11fc0cbf306e\"],[{\"index\":{\"value\":35},\"size\":{\"value\":1}},\"612de1cf-9e9e-464b-be5e-7d6ffaa7c694\"],[{\"index\":{\"value\":37},\"size\":{\"value\":4}},\"fecdad1c-9451-4d2b-a7fd-4aad5550de40\"],[{\"index\":{\"value\":41},\"size\":{\"value\":1}},\"27938428-f93b-429e-9094-2fad6d363c83\"],[{\"index\":{\"value\":42},\"size\":{\"value\":6}},\"edf2376f-4edf-4c43-9111-461dc29cef3d\"],[{\"index\":{\"value\":37},\"size\":{\"value\":11}},\"2d6c6cae-414b-4174-8450-afabd042ae93\"],[{\"index\":{\"value\":28},\"size\":{\"value\":20}},\"7ba77f90-2cf9-4402-a3ab-d6f1e6585d51\"],[{\"index\":{\"value\":53},\"size\":{\"value\":1}},\"20f2ff1b-8932-48e5-ba25-b59d15bf7d58\"],[{\"index\":{\"value\":23},\"size\":{\"value\":31}},\"abd6952c-ab14-4574-8b8d-a089142c37cf\"],[{\"index\":{\"value\":17},\"size\":{\"value\":37}},\"0df7d0e1-a95a-47f3-9ea5-b9e54919db2b\"]]\n{\"ide\":{\"node\":{\"20f2ff1b-8932-48e5-ba25-b59d15bf7d58\":{\"position\":{\"vector\":[-495.9664,199.92416]}},\"2d6c6cae-414b-4174-8450-afabd042ae93\":{\"position\":{\"vector\":[-652.0,87.0]}},\"7ba77f90-2cf9-4402-a3ab-d6f1e6585d51\":{\"position\":{\"vector\":[-584.0,115.0]}},\"32dd808a-0633-4648-994c-691653c6e758\":{\"position\":{\"vector\":[-506.0,116.0]}}}}}"}],"newVersion":"a03427902ca6c19db8721062925a1839133762ddc1498cca25be7b10","oldVersion":"a03427902ca6c19db8721062925a1839133762ddc1498cca25be7b10","path":{"rootId":"d1c75f23-bda3-4738-81fc-d9222d193be3","segments":["src","Main.enso"]}}}}	2619	
{"jsonrpc":"2.0","id":4,"result":{"contextId":"9a09213d-4414-4828-b41e-a32358a6e08b","canModify":{"method":"executionContext/canModify","registerOptions":{"contextId":"9a09213d-4414-4828-b41e-a32358a6e08b"}},"receivesUpdates":{"method":"executionContext/receivesUpdates","registerOptions":{"contextId":"9a09213d-4414-4828-b41e-a32358a6e08b"}}}}	344	
{"jsonrpc":"2.0","id":6,"method":"executionContext/push","params":{"contextId":"9a09213d-4414-4828-b41e-a32358a6e08b","stackItem":{"methodPointer":{"definedOnType":"Main","file":{"rootId":"d1c75f23-bda3-4738-81fc-d9222d193be3","segments":["src","Main.enso"]},"name":"main"},"positionalArgumentsExpressions":[],"thisArgumentExpression":null,"type":"ExplicitCall"}}}	364	
{"jsonrpc":"2.0","id":5,"result":null}	38	
{"jsonrpc":"2.0","id":6,"result":null}	38	
{"jsonrpc":"2.0","id":7,"method":"file/exists","params":{"path":{"rootId":"d1c75f23-bda3-4738-81fc-d9222d193be3","segments":["visualization"]}}}	144	
{"jsonrpc":"2.0","method":"executionContext/expressionValuesComputed","params":{"contextId":"9a09213d-4414-4828-b41e-a32358a6e08b","updates":[{"id":"edf2376f-4edf-4c43-9111-461dc29cef3d","type":null,"shortValue":"UnresolvedSymbol<mulAdd>","methodCall":null}]}}	260	
{"jsonrpc":"2.0","method":"executionContext/expressionValuesComputed","params":{"contextId":"9a09213d-4414-4828-b41e-a32358a6e08b","updates":[{"id":"fecdad1c-9451-4d2b-a7fd-4aad5550de40","type":"Main","shortValue":"Main","methodCall":null}]}}	242	
{"jsonrpc":"2.0","method":"executionContext/expressionValuesComputed","params":{"contextId":"9a09213d-4414-4828-b41e-a32358a6e08b","updates":[{"id":"2d6c6cae-414b-4174-8450-afabd042ae93","type":"Number","shortValue":"5","methodCall":{"file":{"rootId":"d1c75f23-bda3-4738-81fc-d9222d193be3","segments":["src","Main.enso"]},"definedOnType":"Main","name":"mulAdd"}}]}}	365	
{"jsonrpc":"2.0","method":"executionContext/expressionValuesComputed","params":{"contextId":"9a09213d-4414-4828-b41e-a32358a6e08b","updates":[{"id":"7ba77f90-2cf9-4402-a3ab-d6f1e6585d51","type":"Unit","shortValue":"Unit","methodCall":null}]}}	242	
{"jsonrpc":"2.0","method":"executionContext/expressionValuesComputed","params":{"contextId":"9a09213d-4414-4828-b41e-a32358a6e08b","updates":[{"id":"20f2ff1b-8932-48e5-ba25-b59d15bf7d58","type":"Number","shortValue":"6","methodCall":null}]}}	241	
{"jsonrpc":"2.0","method":"executionContext/expressionValuesComputed","params":{"contextId":"9a09213d-4414-4828-b41e-a32358a6e08b","updates":[{"id":"abd6952c-ab14-4574-8b8d-a089142c37cf","type":"Number","shortValue":"6","methodCall":null}]}}	241	
{"jsonrpc":"2.0","method":"executionContext/expressionValuesComputed","params":{"contextId":"9a09213d-4414-4828-b41e-a32358a6e08b","updates":[{"id":"7ba77f90-2cf9-4402-a3ab-d6f1e6585d51","type":"Unit","shortValue":"Unit","methodCall":null}]}}	242	
{"jsonrpc":"2.0","method":"executionContext/expressionValuesComputed","params":{"contextId":"9a09213d-4414-4828-b41e-a32358a6e08b","updates":[{"id":"20f2ff1b-8932-48e5-ba25-b59d15bf7d58","type":"Number","shortValue":"6","methodCall":null}]}}	241	
{"jsonrpc":"2.0","method":"executionContext/expressionValuesComputed","params":{"contextId":"9a09213d-4414-4828-b41e-a32358a6e08b","updates":[{"id":"abd6952c-ab14-4574-8b8d-a089142c37cf","type":"Number","shortValue":"6","methodCall":null}]}}	241	
{"jsonrpc":"2.0","id":7,"result":{"exists":false}}	50	
{"jsonrpc":"2.0","id":8,"method":"executionContext/push","params":{"contextId":"9a09213d-4414-4828-b41e-a32358a6e08b","stackItem":{"expressionId":"2d6c6cae-414b-4174-8450-afabd042ae93","type":"LocalCall"}}}	206	
{"jsonrpc":"2.0","id":8,"result":null}	38	
{"jsonrpc":"2.0","method":"executionContext/expressionValuesComputed","params":{"contextId":"9a09213d-4414-4828-b41e-a32358a6e08b","updates":[{"id":"2c980296-5cb9-42ac-adea-095409fa7c3a","type":"Number","shortValue":"5","methodCall":null}]}}	241	
{"jsonrpc":"2.0","id":9,"method":"executionContext/pop","params":{"contextId":"9a09213d-4414-4828-b41e-a32358a6e08b"}}	118	
{"jsonrpc":"2.0","id":9,"result":null}	38	
{"jsonrpc":"2.0","method":"executionContext/expressionValuesComputed","params":{"contextId":"9a09213d-4414-4828-b41e-a32358a6e08b","updates":[{"id":"7ba77f90-2cf9-4402-a3ab-d6f1e6585d51","type":"Unit","shortValue":"Unit","methodCall":null}]}}	242	
{"jsonrpc":"2.0","method":"executionContext/expressionValuesComputed","params":{"contextId":"9a09213d-4414-4828-b41e-a32358a6e08b","updates":[{"id":"20f2ff1b-8932-48e5-ba25-b59d15bf7d58","type":"Number","shortValue":"6","methodCall":null}]}}	241	
{"jsonrpc":"2.0","method":"executionContext/expressionValuesComputed","params":{"contextId":"9a09213d-4414-4828-b41e-a32358a6e08b","updates":[{"id":"abd6952c-ab14-4574-8b8d-a089142c37cf","type":"Number","shortValue":"6","methodCall":null}]}}	241	
```

Initially, when the execution context is in the main function it gets information about 6 expressions being computed (and actually 3 of them are sent twice — we don't really mind but it is unexpected as well).
Then after entering and leaving node, engine sends `expressionValuesComputed` only for 3 expressions (the ones that were previously reported twice), but other 3 remain uncomputed.

Now, IDE expects to get all the updates every time after entering or leaving node.
Though this could be conceivably a feature, that registry is not invalidated, and the values previously computed are stored by the IDE. But then — the remaining 3 expressions shouldn't be updated. Current behavior doesn't make sense either way.

I expect that:
1. Either all the expressions are reported after changing the stackframe, or we discuss and agree to cache and reuse previously computed values;
2. It should be documented what updates we should expect from the engine after pushing/popping the frame.

### PR DESCRIPTION
**Pull request by 4e6**
### Pull Request Description

close #1029 

PR adds some more corner-case tests updating value notifications handling.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/main/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/main/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guides.
- [x] All documentation and configuration conforms to the [markdown](https://github.com/enso-org/enso/blob/main/docs/style-guide/markdown.md) and [YAML](https://github.com/enso-org/enso/blob/main/docs/style-guide/yaml.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

File: engine/runtime/src/main/java/org/enso/interpreter/instrument/MethodCallsCache.java [hunk 1]
```diff
@@ -20,6 +20,11 @@ public void setExecuted(UUID call) {
     callsExecuted.add(call);
   }
 
+  /** @return the set of executed calls. */
```
Comment (iamrecursion) [inline engine/runtime/src/main/java/org/enso/interpreter/instrument/MethodCallsCache.java:23]: What is an executed call?

Comment (4e6) [reply engine/runtime/src/main/java/org/enso/interpreter/instrument/MethodCallsCache.java:23]: Executed call is a function call that we captured in the instrument during the program execution.

Comment (iamrecursion) [reply engine/runtime/src/main/java/org/enso/interpreter/instrument/MethodCallsCache.java:23]: Is this defined somewhere?

Comment (4e6) [reply engine/runtime/src/main/java/org/enso/interpreter/instrument/MethodCallsCache.java:23]: There is some info in the class definition

```java
/**
 * A storage for executed methods.
 *
 * <p>Tracks the encountered method calls during the invocation of the stack frame. Method calls
 * cache is not preserved between the program invocations.
 */
public final class MethodCallsCache {
```

File: engine/runtime/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala [hunk 2]
```diff
@@ -1324,7 +1324,7 @@ class RuntimeServerTest
           Vector(
             Api.ExpressionValueUpdate(
               idMainA,
-              Some("Number"),
+              None,
```
Comment (kustosz) [inline engine/runtime/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala]: Why do these become `None`? Does IDE not want to get the type here? Doesn't sound good

Comment (kustosz) [reply engine/runtime/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala]: If these fields become independent, it's probably better to split the message in two? The semantics of `there's a method call but no type` are unclear to me.

**Review by kustosz — APPROVED** [review]
Approval conditional on the results of the discussion started in my inline comment.

File: engine/runtime/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala (hunk 2 above)
Comment (4e6) [reply engine/runtime/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala]: There's two optional fields for the _type_ and _method pointer_ in the `ExpressionValueUpdate` message because IDE wants to see the updates only when the stuff has changed. In this case, we don't send the type because only the method pointer has changed.

PR just updates the notification logic to be more accurate, i.e. previously we always send the type whenever the method pointer changed, and now we don't

Comment (kustosz) [reply engine/runtime/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala]: I see. If that's what the IDE expects really, land it :)

Comment (farmaazon) [reply engine/runtime/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala:1327]: Hold a second. As I know (and I asked @mwu-tow ) as well, we don't want to None here.

Comment (farmaazon) [reply engine/runtime/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala]: Hold a second. As I know (and I asked @mwu-tow ) as well, we don't want to None here.

**Review by farmaazon — CHANGES_REQUESTED** [review]

File: engine/runtime/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala (hunk 2 above)
Comment (farmaazon) [reply engine/runtime/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala]: The issue was that we don't want to receive update **packet** when there was no change. But if the update is fired it should contain all the information.

We treat "None" here as "There is no known type yet", no "The type was not changed from last update".

Maybe again I miss or forget some agreement, but I don't see them in my discord history. If there is some issue comment with request for _skipping unchanged fields_, could you link it here?

Comment (mwu-tow) [reply engine/runtime/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala]: @4e6 IDE doesn't want spurious updates, but it was about not emitting the same updates multiple times, as expression is re-evaluated without change in the code or execution context. 

Emitting an update with type `None` would actually mean "we don't know what the type is anymore" and is the desired behavior.

If we wanted to really optimize for a case where type remains the same but the method pointer changes, we should introduce separate notification for them. (or have `type` field become a 3-variant enum: `Unchanged`, `Unknown`, `Updated(typename)` — so we could handle such partial updates) — but both of these feel like an overkill.

Basically we want to have as few `ExpressionValueUpdate` as possible but not less (i.e. have them after changing frame or its properties changed). We never requested "partial" updates.

Sorry if I haven't made this clear properly before.

**Review by mwu-tow — CHANGES_REQUESTED** [review]

**Comment by 4e6** [comment]
@mwu-tow 

> Emitting an update with type None would actually mean "we don't know what the type is anymore" and is the desired behavior.

Runtime knows about all its types, they are limited to this set (@kustosz correct me if I'm wrong):

https://github.com/enso-org/enso/blob/5fcd80d4d08c5040f5a39e42bd4ecba0ff840cb0/engine/runtime/src/main/java/org/enso/interpreter/runtime/type/Types.java#L27-L37

So, it can only send `None` type if the value has not changed.

**Comment by farmaazon** [comment]
@4e6 
> 
> Runtime knows about all its types, they are limited to this set (@kustosz correct me if I'm wrong):
> 
> https://github.com/enso-org/enso/blob/5fcd80d4d08c5040f5a39e42bd4ecba0ff840cb0/engine/runtime/src/main/java/org/enso/interpreter/runtime/type/Types.java#L27-L37
> 
> So, it can only send `None` type if the value has not changed.

But the other field: methodCall in turn has not distinction between "I haven't changed" from "There is not method called in this expression", and I'd prefer the unified approach.

**Comment by mwu-tow** [comment]
@4e6 
So to clarify:
1) we will never get `None`, unless this was already computed in this frame;
2) after changing the frame we won't get `None` but the properly filled type?

Still, it feels like a strange API design, because, if I understand correctly, we would have structure with two optional fields and totally different `None` semantics for them.  (for type it is "no change" and for method it is "not a call"). 
If this is intended and preferred design for the engine, IDE can adjust to this (change will be really simple). But the `None` meanings for fields needs to be part of the API documentation.

Also, documentation should include information about when these updates are emitted (as we are requesting changes and want to rely on this being implemented that way — so it should be part of documented API behavior)

**Comment by kustosz** [comment]
I agree the proposed design is strange (and the reason why I opposed at the beginning). Also what happens when the type changes but the method call does not? `null` already means "no method call", not "reuse method call". Therefore if we really wan't to have this partial-update behavior we need a tri-state type for each field, or separate the message into different messages.
Also thinking about the future extensions, when we have the typechecker, it will become even stranger with the proposed semantics – the typechecker will only send a type with no value and possibly a method call (but with no guarantee), then the runtime would refine the original type and update the method call, with caching on top of all that. Abusing null semantics will quickly become unwieldy. Therefore I vote we redesign the API or always send both values.

**Comment by 4e6** [comment]
Ok, we have a notification with two fields, `type` and `methodPointer`. What if we model "field not changed" state by omitting the key-value pair from the JSON response. More precisely:

- if the `"key": "Value"` pair is present in JSON response, it means that the type or method pointer is changed.
  - if the `"type"` value is `null`, that the type of expression is unknown (shouldn't really happen)
  - if the `"methodPointer": null`, means there's no method call associated with the expression
- if the `"key": "Value"` pair is omitted from JSON response, it means that the type or method pointer has not changed

**Comment by 4e6** [comment]
Alternatively, we can make all fields mandatory. So the update notification will mean - the expression has changed, here is its new state (type and method pointer). This approach should be more simple.

**Comment by kustosz** [comment]
The first proposal is effectively modelling a tri-state but with more error-prone syntax. The second proposal is "leave things as they were before" and I don't mind, so @mwu-tow ?

**Comment by mwu-tow** [comment]
@kustosz 
> The first proposal is effectively modelling a tri-state but with more error-prone syntax. The second proposal is "leave things as they were before" and I don't mind, so @mwu-tow ?

Discerning between `null` / `value` / [lack of field] might be troublesome. I'm for leaving things as they were, i.e. having fields set whenever possible.
The field values are IMO light-weight enough to not worry about sending them even if they are the same.

**Comment by 4e6** [comment]
@farmaazon @mwu-tow Ok, I updated the implementation to send the full update containing both, new type and method pointer, when one of them is changed.

**Review by mwu-tow — APPROVED** [review]

**Review by farmaazon — APPROVED** [review]
